src/server/pps/server/api_server.go

```diff
@@ -145,26 +150,87 @@ func (a *apiServer) CreateJob(ctx context.Context, request *ppsclient.CreateJobR
 		}
 	}
 
+	repoToFromCommit := make(map[string]*pfsclient.Commit)
+	if parentJobInfo != nil {
+		for _, jobInput := range parentJobInfo.Inputs {
+			if !jobInput.Reduce {
+				// input isn't being reduced, do it incrementally
+				repoToFromCommit[jobInput.Commit.Repo.Name] = jobInput.Commit
+			}
+		}
+	}
+
+	var viableFilterShards []*pfsclient.Shard
+	for i := 0; i < int(request.Parallelism); i++ {
+	CheckInputs:
+		for _, jobInput := range request.Inputs {
+			listFileRequest := &pfsclient.ListFileRequest{
+				File: &pfsclient.File{
+					Commit: jobInput.Commit,
+					Path:   "", // the root directory
+				},
+				FromCommit: repoToFromCommit[jobInput.Commit.Repo.Name],
+				Shard: &pfsclient.Shard{
+					FileModulus:  1,
+					BlockModulus: 1,
+				},
+				Recurse: true,
+			}
+			if jobInput.Reduce {
+				listFileRequest.Shard.FileNumber = uint64(i)
+				listFileRequest.Shard.FileModulus = request.Parallelism
+			} else {
+				listFileRequest.Shard.BlockNumber = uint64(i)
+				listFileRequest.Shard.BlockModulus = request.Parallelism
+			}
+			fileInfos, err := pfsAPIClient.ListFile(ctx, listFileRequest)
+			if err != nil {
+				return nil, err
+			}
+			for _, fileInfo := range fileInfos.FileInfo {
+				if fileInfo.SizeBytes > 0 {
+					viableFilterShards = append(viableFilterShards, listFileRequest.Shard)
+					break CheckInputs
+				}
+			}
+		}
+	}
+
+	if len(viableFilterShards) == 0 && len(request.Inputs) > 0 {
+		return nil, ErrEmptyInput
+	}
+
```

Comment:
I got a little bit confused trying to understand this logic, how would you feel about calling `viableFilterShards` `nonEmptyFilterShards` instead? I think that would have helped me get it sooner.